include/TSettings.h

```diff
@@ -0,0 +1,321 @@
+    auto format(ComposedKey key, format_context& ctx) const;
+};
+
+inline auto fmt::formatter<ComposedKey>::format(ComposedKey key, fmt::format_context& ctx) const {
+    std::string key_metadata = fmt::format("{}::{}", key.Category, key.Key);
+    return formatter<std::string>::format(key_metadata, ctx);
+}
+
+namespace std {
+template <>
+class hash<ComposedKey> {
+public:
+    std::uint64_t operator()(const ComposedKey& key) const {
+        std::hash<std::string> hash_fn;
+        return hash_fn(key.Category + key.Key);
+    }
+};
+}
+
+struct Settings {
+    using SettingsTypeVariant = std::variant<std::string, bool, int>;
+
+    enum Key {
+        // Keys that correspond to the keys set in TOML
+        // Keys have their TOML section name as prefix
+
+        // [Misc]
+        Misc_SendErrorsShowMessage,
+        Misc_SendErrors,
+        Misc_ImScaredOfUpdates,
+
+        // [General]
+        General_Description,
+        General_Tags,
+        General_MaxPlayers,
+        General_Name,
+        General_Map,
+        General_AuthKey,
+        General_Private,
+        General_Port,
+        General_MaxCars,
+        General_LogChat,
+        General_ResourceFolder,
+        General_Debug
+    };
+
+    Sync<std::unordered_map<Key, SettingsTypeVariant>> SettingsMap = std::unordered_map<Key, SettingsTypeVariant> {
+        { General_Description, std::string("BeamMP Default Description") },
+        { General_Tags, std::string("Freeroam") },
+        { General_MaxPlayers, 8 },
+        { General_Name, "BeamMP Server" },
+        { General_Map, "/levels/gridmap_v2/info.json" },
+        { General_AuthKey, "" },
+        { General_Private, true },
+        { General_Port, 30814 },
+        { General_MaxCars, 1 },
+        { General_LogChat, true },
+        { General_ResourceFolder, "Resources" },
+        { General_Debug, false },
+        { Misc_SendErrorsShowMessage, true },
+        { Misc_SendErrors, true },
+        { Misc_ImScaredOfUpdates, true }
+    };
+
+    enum SettingsAccessMask {
+        read, // Value can be read from console
+        write, // Value can be read and written to from console
+        noaccess // Value is inaccessible from console (no read OR write)
+    };
+
+    using SettingsAccessControl = std::pair<
+        Key, // The Key's corresponding enum encoding
+        SettingsAccessMask // Console read/write permissions
+        >;
+
+    Sync<std::unordered_map<ComposedKey, SettingsAccessControl>> InputAccessMapping = std::unordered_map<ComposedKey, SettingsAccessControl> {
+        { { "General", "Description" }, { General_Description, write } },
+        { { "General", "Tags" }, { General_Tags, write } },
+        { { "General", "MaxPlayers" }, { General_MaxPlayers, write } },
+        { { "General", "Name" }, { General_Name, write } },
+        { { "General", "Map" }, { General_Map, read } },
+        { { "General", "AuthKey" }, { General_AuthKey, noaccess } },
```

Comment:
write perms for changing it

Reply:
wdym, all of these should get write access from runtime?